src/tools/crossgen2/Common/TypeSystem/Common/ExplicitLayoutValidator.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Internal.TypeSystem;
+using System;
+using System.Net;
+
+namespace Internal.TypeSystem
+{
+    public class ExplicitLayoutValidator
+    {
+        private enum FieldLayoutTag : byte
+        {
+            Empty,
+            NonORef,
+            ORef,
+        }
+
+        private readonly int _pointerSize;
+
+        private readonly FieldLayoutTag[] _fieldLayout;
+
+        private readonly MetadataType _typeBeingValidated;
+
+        private ExplicitLayoutValidator(MetadataType type, int typeSizeInBytes)
+        {
+            _typeBeingValidated = type;
+            _pointerSize = type.Context.Target.PointerSize;
+            _fieldLayout = new FieldLayoutTag[typeSizeInBytes];
+        }
+
+        public static void Validate(MetadataType type, ComputedInstanceFieldLayout layout)
+        {
+            ExplicitLayoutValidator validator = new ExplicitLayoutValidator(type, layout.ByteCountUnaligned.AsInt);
+            foreach (FieldAndOffset fieldAndOffset in layout.Offsets)
+            {
+                validator.AddToFieldLayout(fieldAndOffset.Offset.AsInt, fieldAndOffset.Field.FieldType);
+            }
+        }
+
+        private void AddToFieldLayout(int offset, TypeDesc fieldType)
+        {
+            if (fieldType.IsGCPointer)
+            {
+                if (offset % _pointerSize != 0)
+                {
+                    // Misaligned ORef
+                    ThrowFieldLayoutError(offset);
+                }
+                SetFieldLayout(offset, _pointerSize, FieldLayoutTag.ORef);
+            }
+            else if (fieldType.IsValueType)
+            {
+                MetadataType mdType = (MetadataType)fieldType;
```

Comment:
If the field is a ByRef-like struct, it also needs to be aligned at pointer boundary. This should throw:

```csharp
[StructLayout(LayoutKind.Explicit)]
ref struct Program
{
    [FieldOffset(2)] Span<int> f;

    static unsafe void Main()
    {
    }
}
```

Reply:
Added alignment check for IsByRefLike within the IsValueType conditional block.